src/app/clusters/fan-control-server/fan-control.h

```diff
@@ -0,0 +1,393 @@
+
+    /**
+     * Internal airflow direction write handler function.
+     */
+    chip::Protocols::InteractionModel::Status handleAirflowDirectionWrite(AttributeValueDecoder & aDecoder);
+
+    /**
+     * Internal step command handler function.
+     */
+    void handleStep(HandlerContext & ctx, const Commands::Step::DecodableType & commandData);
+};
+
+class Delegate
+{
+public:
+    virtual ~Delegate() = default;
+
+    /**
+     * This init function will be called during the FanControl server initialization after the Instance information has been
+     * validated and the Instance has been registered. This can be used to initialise app logic if required.
+     */
+    virtual CHIP_ERROR Init() { return CHIP_NO_ERROR; }
+
+    /**
+     * This function will be called when the FanMode attribute is changed
+     * @param newMode The new FanMode value.
+     * @return Returns a CHIP_NO_ERROR if there was no error.
+     */
+    virtual CHIP_ERROR HandleFanModeChange(FanModeEnum newMode) = 0;
+
+    /**
+     * This function will be called when the FanMode attribute is changed to Off. The server will change the SpeedSetting and
+     * PercentSetting in this case so are included in the callback.
+     * @param newPercentSetting The new PercentSetting value.
+     * @param newSpeedSetting The new SpeedSetting value (can be ignored if the feature is not supported)
+     * @return Returns a CHIP_NO_ERROR if there was no error.
+     *
+     */
+    virtual CHIP_ERROR HandleFanModeOff(DataModel::Nullable<Percent> newPercentSetting,
+                                        DataModel::Nullable<uint8_t> newSpeedSetting) = 0;
+
+    /**
+     * This function will be called when the FanMode attribute is changed to Auto. The server will change the SpeedSetting and
+     * PercentSetting in this case so are included in the callback.
+     * @param newPercentSetting The new PercentSetting value.
+     * @param newSpeedSetting The new SpeedSetting value (can be ignored if the feature is not supported)
+     * @return Returns a CHIP_NO_ERROR if there was no error.
+     */
+    virtual CHIP_ERROR HandleFanModeAuto(DataModel::Nullable<Percent> newPercentSetting,
+                                         DataModel::Nullable<uint8_t> newSpeedSetting)
+    {
+        return CHIP_ERROR_NOT_IMPLEMENTED;
+    }
+
+    /**
+     * This function will be called when the PercentSetting or SpeedSetting attributes are changed.
+     * @param newPercentSetting The new PercentSetting value.
+     * @param newSpeedSetting The new SpeedSetting value (can be ignored if the feature is not supported)
+     * @return Returns a CHIP_NO_ERROR if there was no error.
+     */
+    virtual CHIP_ERROR HandlePercentSpeedSettingChange(DataModel::Nullable<Percent> newPercentSetting,
+                                                       DataModel::Nullable<uint8_t> newSpeedSetting) = 0;
+
+    /**
+     * This function will be called when the RockSetting attribute is changed.
```

Comment:
This, and the other callbacks, need to be very clear that they will be called when the relevant attributes are changed by _Matter interactions_, but not when the attributes are changed via calls to the cluster's public API, right?

It might be better to use "written to" instead of "changed" to make that clearer, but might also just need a longer description.

Also should document whether the values passed to these functions will have already been constraint-checked based on the spec constraints (yes, they will be if the comments above are fixed).  The delegate in all-clusters-app could then stop doing all sorts of checks it does right now.